File: Data Structures & Algorithms/two-integer-sum/submission-9.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @return {number[]}
     */
    twoSum(nums, target) {
        const visitedMap = new Map();
        for (let i = 0; i < nums.length; i++) {
            const complement = target - nums[i];
            if (visitedMap.has(complement)) {
                return [visitedMap.get(complement), i];
            }

            visitedMap.set(nums[i], i);
        }
    }
}

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-9.java
class Solution {
    Boolean[][] dp;
    public boolean canPartition(int[] nums) {
        int sum  = 0;
        for (int num : nums)
            sum += num;
        if (sum % 2 != 0)
            return false;

        sum /= 2;
        int n = nums.length;
        dp = new Boolean[n][sum + 1];
        return canPartition(nums, 0, sum);

    }

    private boolean canPartition(int[] nums, int index, int target) {
        if (index >= nums.length || target < 0)
            return false;
        
        if (dp[index][target] != null)
            return dp[index][target];

        if (target == 0) {
            dp[index][target] = true;
            return dp[index][target];
        }
        
        dp[index][target] = canPartition(nums, index + 1, target - nums[index]) || canPartition(nums, index + 1, target);
        return dp[index][target];
    }


}
